File: test/features/settings/health_connect_sources_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:openvitals/features/settings/application/health_connect_sources.dart';

void main() {
  DateTime t(int day) => DateTime.utc(2026, 7, day);

  group('aggregateHealthConnectSources', () {
    test('folds observations per package, counting and dating them', () {
      final sources = aggregateHealthConnectSources({
        'heart rate': [
          ('com.sec.android.app.shealth', t(1)),
          ('com.sec.android.app.shealth', t(3)),
          ('tech.mmarca.openvitals', t(2)),
        ],
        'sleep': [
          ('com.sec.android.app.shealth', t(4)),
        ],
      });

      expect(sources.length, 2);
      // Sorted most-recent-contribution first: Samsung (day 4) before us (day 2).
      expect(sources.first.package, 'com.sec.android.app.shealth');
      expect(sources.first.recordCount, 3);
      expect(sources.first.lastSeen, t(4));
      expect(sources.first.metrics, {'heart rate', 'sleep'});
      expect(sources.first.displayName, 'Samsung Health');

      expect(sources.last.package, 'tech.mmarca.openvitals');
      expect(sources.last.metrics, {'heart rate'});
      expect(sources.last.displayName, 'OpenVitals (this app)');
    });

    test('blank sources collapse to a single "unknown" entry', () {
      final sources = aggregateHealthConnectSources({
        'heart rate': [('', t(1)), ('   ', t(2))],
      });

      expect(sources.single.package, 'unknown');
      expect(sources.single.recordCount, 2);
      expect(sources.single.displayName, 'Unknown source');
    });

    test('an empty read yields no sources', () {
      expect(
        aggregateHealthConnectSources({'heart rate': const [], 'sleep': const []}),
        isEmpty,
      );
    });

    test('an unknown package keeps its raw name', () {
      final sources = aggregateHealthConnectSources({
        'sleep': [('com.example.watch', t(1))],
      });
      expect(sources.single.displayName, 'com.example.watch');
    });
  });
}

File: app/src/test/kotlin/tech/mmarca/openvitals/features/settings/HealthConnectSourcesTest.kt
package tech.mmarca.openvitals.features.settings

import java.time.Instant
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class HealthConnectSourcesTest {

    private val t1 = Instant.parse("2026-07-01T10:00:00Z")
    private val t2 = Instant.parse("2026-07-02T10:00:00Z")
    private val t3 = Instant.parse("2026-07-03T10:00:00Z")

    @Test
    fun `aggregates counts and metrics per package`() {
        val sources = aggregateHealthConnectSources(
            mapOf(
                "heart rate" to listOf(
                    "com.sec.android.app.shealth" to t1,
                    "com.sec.android.app.shealth" to t2,
                ),
                "sleep" to listOf(
                    "com.sec.android.app.shealth" to t3,
                ),
            ),
        )

        assertEquals(1, sources.size)
        val samsung = sources.single()
        assertEquals(3, samsung.recordCount)
        assertEquals(setOf("heart rate", "sleep"), samsung.metrics)
        assertEquals(t3, samsung.lastSeen)
        assertEquals("Samsung Health", samsung.displayName)
    }

    @Test
    fun `sorts most recent contributor first`() {
        val sources = aggregateHealthConnectSources(
            mapOf(
                "heart rate" to listOf(
                    "old.app" to t1,
                    "new.app" to t3,
                ),
            ),
        )

        assertEquals(listOf("new.app", "old.app"), sources.map { it.packageName })
        // Unrecognized packages fall back to the raw name.
        assertEquals("new.app", sources.first().displayName)
    }

    @Test
    fun `blank sources fold into the unknown bucket`() {
        val sources = aggregateHealthConnectSources(
            mapOf(
                "heart rate" to listOf(
                    "" to t1,
                    "   " to t2,
                ),
            ),
        )

        assertEquals(1, sources.size)
        assertEquals(HealthConnectSource.UNKNOWN_PACKAGE, sources.single().packageName)
        assertEquals("Unknown source", sources.single().displayName)
        assertEquals(2, sources.single().recordCount)
    }

    @Test
    fun `empty input yields an empty list`() {
        assertTrue(aggregateHealthConnectSources(emptyMap()).isEmpty())
        assertTrue(aggregateHealthConnectSources(mapOf("sleep" to emptyList())).isEmpty())
    }
}
